tests/unit/s2n_psk_test.c

```diff
@@ -115,6 +115,48 @@ int main(int argc, char **argv)
         EXPECT_BYTEARRAY_EQUAL(psk.secret.data, TEST_VALUE_2, sizeof(TEST_VALUE_2));
     }
 
+    /* Test s2n_psk_clone */
+    {
+        const uint8_t test_bad_value[] = "wrong";
+        const uint8_t test_identity[] = "identity";
+        const uint8_t test_secret[] = "secret";
+        const uint8_t test_early_secret[] = "early_secret";
+        const s2n_hmac_algorithm test_hmac = S2N_HMAC_SHA384;
+
+        struct s2n_psk *original = s2n_external_psk_new();
+        EXPECT_NOT_NULL(original);
+        EXPECT_SUCCESS(s2n_psk_set_identity(original, test_identity, sizeof(test_identity)));
+        EXPECT_SUCCESS(s2n_psk_set_secret(original, test_secret, sizeof(test_secret)));
+        EXPECT_SUCCESS(s2n_alloc(&original->early_secret, sizeof(test_early_secret)));
+        EXPECT_MEMCPY_SUCCESS(original->early_secret.data, test_early_secret, original->early_secret.size);
+        original->hmac_alg = test_hmac;
+
+        DEFER_CLEANUP(struct s2n_psk *clone = s2n_external_psk_new(), s2n_psk_free);
+        EXPECT_SUCCESS(s2n_psk_set_identity(clone, test_bad_value, sizeof(test_bad_value)));
+        EXPECT_SUCCESS(s2n_psk_set_secret(clone, test_bad_value, sizeof(test_bad_value)));
+        EXPECT_NOT_NULL(clone);
+
+        EXPECT_OK(s2n_psk_clone(clone, original));
+
+        /* Free the original to ensure they share no memory */
+        EXPECT_SUCCESS(s2n_psk_free(&original));
+
+        /* existing identity is replaced by original's identity */
+        EXPECT_EQUAL(clone->identity.size, sizeof(test_identity));
+        EXPECT_BYTEARRAY_EQUAL(clone->identity.data, test_identity, sizeof(test_identity));
+
+        /* new secret is replaced by original's secret */
+        EXPECT_EQUAL(clone->secret.size, sizeof(test_secret));
+        EXPECT_BYTEARRAY_EQUAL(clone->secret.data, test_secret, sizeof(test_secret));
+
+        /* early secret is allocated for original's early secret */
+        EXPECT_EQUAL(clone->early_secret.size, sizeof(test_early_secret));
+        EXPECT_BYTEARRAY_EQUAL(clone->early_secret.data, test_early_secret, sizeof(test_early_secret));
+
+        /* other values are copied */
+        EXPECT_EQUAL(clone->hmac_alg, test_hmac);
```

Comment:
We should assert the `type` and  `obfuscated_ticket_age` are also copied. We are also missing checks for `early_data_config` being cloned here.

Reply:
early_data_config being cloned is handled in the early_data_config tests.

Reply:
Yes, but `s2n_psk_clone` contract includes that `early_data_config` is also copied and we should assert that here.

Reply:
We would still need to assert `type` and `obfuscated_ticket_age` here?